include/deal.II/non_matching/mapping_info.h

```diff
@@ -885,32 +997,97 @@ namespace NonMatching
             cell_level_and_indices[cell_index] = {cell->level(), cell->index()};
           }
 
+        const auto &quadrature = quadrature_vector[cell_index];
+        const bool  empty      = quadrature.empty();
+
         // store unit points
         const unsigned int n_q_points = compute_n_q_points<VectorizedArrayType>(
           n_q_points_unvectorized[cell_index]);
         store_unit_points(unit_points_index[cell_index],
                           n_q_points,
                           n_q_points_unvectorized[cell_index],
-                          quadrature_vector[cell_index].get_points());
+                          quadrature.get_points());
 
         // compute mapping data
-        internal::ComputeMappingDataHelper<dim, spacedim>::
-          compute_mapping_data_for_quadrature(mapping,
-                                              update_flags_mapping,
-                                              cell,
-                                              cell_similarity,
-                                              quadrature_vector[cell_index],
-                                              internal_mapping_data,
-                                              mapping_data);
+        compute_mapping_data(cell, quadrature, mapping_data);
 
         // store mapping data
         const unsigned int n_q_points_data =
           compute_n_q_points<Number>(n_q_points_unvectorized[cell_index]);
+
+        // check for cartesian/affine cell
+        if (!empty &&
+            update_flags_mapping & UpdateFlags::update_inverse_jacobians)
+          {
+            cell_type.push_back(
+              internal::compute_geometry_type(cell->diameter(),
+                                              mapping_data.inverse_jacobians));
+          }
+        else
+          cell_type.push_back(
+            dealii::internal::MatrixFreeFunctions::GeometryType::general);
+
+        if (cell_index > 0)
+          {
+            // check if current and last cell are affine
+            const bool affine_cells =
+              cell_type[cell_index] <=
+                dealii::internal::MatrixFreeFunctions::affine &&
+              cell_type[cell_index - 1] <=
+                dealii::internal::MatrixFreeFunctions::affine;
+
+            // create a comparator to compare inverse Jacobian of current
+            // and last cell
+            FloatingPointComparator<double> comparator(
+              1e4 / cell->diameter() * std::numeric_limits<double>::epsilon() *
+              1024.);
+
+            // we can only compare if current and last cell have at least
+            // one quadrature point and both cells are at least affine
```

Comment:
Are you planing to implement something more sophisticated?

Reply:
Yes I do, but not at the moment. The next step will be to detect local refinements.

Reply:
For the time when we start thinking about extending the code here, I suggest we try to unify this with the regular `MatrixFreeFunctions::MappingInfo` because most of the functionality there, in particular https://github.com/dealii/dealii/blob/605a6268901ad87cec3d828e742725723298051b/include/deal.II/matrix_free/mapping_info.templates.h#L1066-L1073 which has an `std::map` to detect also long-range dependencies, e.g. in extruded meshes.